tools/spin-ec2/aws.go

```diff
@@ -0,0 +1,153 @@
+	"strings"
+)
+
+var versionRegex = regexp.MustCompile(`v[0-9]+\.[0-9]+\.[0-9]+`)
+
+// getAWSInstances return all the canary instance names mapped on ids.
+func getAWSInstances() (map[string]string, error) {
+	cmd := exec.Command("aws", []string{
+		"ec2",
+		"describe-instances",
+		"--output", "text",
+		"--filters", `Name=tag:Name,Values="` + hostPrefix + `:*"`, `Name=instance-state-name,Values=running`,
+		"--query", "Reservations[*].Instances[*].[Tags[?Key==`Name`], InstanceId]",
+	}...)
+
+	var out bytes.Buffer
+	var outErr bytes.Buffer
+	cmd.Stdout = &out
+	cmd.Stderr = &outErr
+
+	err := cmd.Run()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get aws instances, error: %s", outErr.String())
+	}
+
+	return parseAWSInstances(out.String())
+}
+
+// parseAWSInstances parses the ec2 describe-instances response.
+func parseAWSInstances(response string) (map[string]string, error) {
+	result := make(map[string]string)
+	split := strings.Split(response, "\n")
+
+	for i := 0; i < len(split)-1; i += 2 {
+		instanceID := strings.TrimSpace(split[i])
+
+		if !strings.HasPrefix(split[i+1], "Name") {
+			return nil, fmt.Errorf("unexpected line received for instanceName: %s", split[i+1])
+		}
+		instanceName := strings.TrimSpace(strings.TrimPrefix(split[i+1], "Name"))
+
+		result[instanceID] = instanceName
+	}
+
+	return result, nil
+}
+
+// pruneInstances will terminate all instances except the ones that have
+// latest 2 infra-agent version installed.
+func pruneInstances(instances map[string]string, dryRun bool) error {
+
+	idsToTerminate, err := getInstancesToPrune(instances)
+	if err != nil {
+		return err
+	}
+
+	if len(idsToTerminate) == 0 {
+		return nil
+	}
+
+	for _, id := range idsToTerminate {
+		fmt.Println(colorizeRed(
+			fmt.Sprintf("[DryRun: '%t'] Removing the following instance: %s -> %s",
+				dryRun, id, instances[id])))
+
+	}
+
+	args := []string{
+		"ec2",
+		"terminate-instances",
+		"--instance-ids",
+	}
+	args = append(args, idsToTerminate...)
+
+	if !dryRun {
+		cmd := exec.Command("aws", args...)
+
+		var out bytes.Buffer
+		var outErr bytes.Buffer
+		cmd.Stdout = &out
+		cmd.Stderr = &outErr
+
+		err := cmd.Run()
+		if err != nil {
+			return fmt.Errorf("failed to terminate aws instances, error: %s", outErr.String())
+		}
+
+		fmt.Println(out.String())
+	}
+
+	return nil
+}
+
+// getInstancesToPrune will filter the instances that should be terminated.
+func getInstancesToPrune(instances map[string]string) ([]string, error) {
+	versions, err := detectVersions(instances)
+	if err != nil {
+		return nil, err
+	}
+
+	var idsToRemove []string
+
+	for _, versionToRemove := range getVersionsToRemove(versions) {
+		for id, name := range instances {
+			if !strings.Contains(name, versionToRemove) {
+				continue
+			}
+			idsToRemove = append(idsToRemove, id)
+		}
+	}
+
+	return idsToRemove, nil
+}
+
+// getVersionsToRemove keeps the latest 2 version of infra agent.
+func getVersionsToRemove(versions []string) []string {
```

Comment:
What if we have only 1 or 2 versions in canaries yet? Will it remove them?

Reply:
it returns an empty array if len(versions) is not > 2